app/Judite/Models/Student.php

```diff
@@ -112,11 +129,23 @@ public function isEnrolledInCourse(Course $course): bool
      *
      * @param \App\Judite\Models\Course $course
      *
+     * @throws \App\Exceptions\UserIsNotEnrolledInCourseException|\App\Exceptions\EnrollmentCannotBeDeleted
+     *
      * @return bool
      */
     public function unenroll(Course $course): bool
     {
-        return $this->enrollments()->where('course_id', $course->id)->delete();
+        if (! $this->isEnrolledInCourse($course)) {
+            throw new UserIsNotEnrolledInCourseException($course);
+        }
+
+        $enrollment = $this->getEnrollmentWithShiftByCourse($course);
```

Comment:
Correct me if I'm wrong, but I really think that we are only hit the database once, please check the method `getEnrollmentWithShiftByCourse(Course $course)`. With the delete we hit the database twice.
This method is used to validate if the a use can delete an enrollment. A student can delete a enrollment only if the enrollment has no associated shift.

Reply:
The condition hits the database, also.

`if(! $this->isEnrolledInCourse($course))`

Reply:
The method will be useful in my new commit to create a method to check if a enrollment is delectable.

Reply:
Still, IMO, we do not need a new method, otherwise it is very easy to hit the database multiple times.

What about this ?


```php

    public function unenroll(Course $course): bool
    {
        $enrollment = $this->enrollments()->where('course_id', $course->id)->first(); // First database hit
        
        if (is_null($enrollment)) {
            throw new UserIsNotEnrolledInCourseException($course);
        }
        
        if (! $enrollment->hasShift()) {
            throw new EnrollmentCannotBeDeleted($enrollment, 'The enrollment cannot be deleted because it as an associated shift.');
        }

        return $this->enrollments()->whereCourseId($course->id)->delete(); // Second database hit
    }

```